packages/@glimmer/program/lib/program.ts

```diff
@@ -71,14 +74,26 @@ export class Heap implements CompileTimeHeap {
       this.table = table;
       this.offset = this.heap.length;
       this.handle = handle;
+      this.pageSize = this.heap.length;
     } else {
-      this.heap = new Uint16Array(0x100000);
+      this.heap = new Uint16Array(PAGE_SIZE);
       this.table = [];
     }
   }
 
   push(item: number): void {
     this.heap[this.offset++] = item;
+    this.sizeCheck();
+  }
+
+  private sizeCheck() {
+    this.pageSize++;
+    if (this.pageSize >= PAGE_SIZE) {
+      let heap = slice(this.heap, 0, this.offset);
```

Comment:
I don't understand why the copy of the heap is necessary, is it not enough to keep a reference to the old heap in a local variable?